sourcetool/pkg/sourcetool/implementation.go

```diff
@@ -54,3 +102,325 @@ func (impl *defaultToolImplementation) GetActiveControls(opts *Options) (slsa.Co
+}
+
+func (impl *defaultToolImplementation) CheckPolicyFork(opts *Options) error {
+	policyOrg, policyRepo, ok := strings.Cut(opts.PolicyRepo, "/")
+	if !ok || policyRepo == "" {
+		return fmt.Errorf("unable to parse policy repository slug")
+	}
+	userForkOrg := opts.UserForkOrg
+	userForkRepo := policyRepo // For now we only support forks with the same name
+
+	// Check the user has a fork of the slsa repo
+	if err := kgithub.VerifyFork(
+		fmt.Sprintf("slsa-source-policy-%d", time.Now().Unix()), userForkOrg, userForkRepo, policyOrg, policyRepo,
+	); err != nil {
+		return fmt.Errorf(
+			"while checking fork of %s/%s in %s: %w ",
+			policyOrg, policyRepo, userForkOrg, err,
+		)
+	}
+	return nil
+}
+
+// CreatePolicyPR creates a pull request to push the policy
+func (impl *defaultToolImplementation) CreatePolicyPR(opts *Options) error {
+	// Branchname to be created on the user's fork
+	branchname := fmt.Sprintf("slsa-source-policy-%d", time.Now().Unix())
+
+	gh := kgithub.New()
+	policyOrg, policyRepo, ok := strings.Cut(opts.PolicyRepo, "/")
+	if !ok || policyRepo == "" {
+		return fmt.Errorf("unable to parse policy repository slug")
+	}
+
+	userForkOrg := opts.UserForkOrg
+	userForkRepo := policyRepo // For now we only support forks with the same name
+
+	// Clone the slsa repo
+	gitCloneOpts := &gogit.CloneOptions{Depth: 1}
+	repo, err := kgithub.PrepareFork(
+		branchname, policyOrg, policyRepo,
+		userForkOrg, userForkRepo,
+		opts.UseSSH, opts.UpdateRepo, gitCloneOpts,
+	)
+	if err != nil {
+		return fmt.Errorf("while preparing Slsa Source fork: %w", err)
+	}
+
+	defer func() {
+		repo.Cleanup() //nolint:errcheck,gosec
+	}()
+
+	// Create the policy in the local clone
+	ghc := ghcontrol.NewGhConnection(opts.Owner, opts.Repo, opts.Branch).WithAuthToken(os.Getenv(tokenVar))
+	outpath, err := policy.CreateLocalPolicy(context.Background(), ghc, repo.Dir())
+	if err != nil {
+		return fmt.Errorf("creating local policy: %w", err)
+	}
+
+	// add the modified manifest to staging
+	logrus.Debugf("Adding %s to staging area", outpath)
+	if err := repo.Add(strings.TrimPrefix(strings.TrimPrefix(outpath, repo.Dir()), "/")); err != nil {
+		return fmt.Errorf("adding new policy file to staging area: %w", err)
+	}
+
+	commitMessage := fmt.Sprintf("Add %s/%s SLSA Source policy file", opts.Owner, opts.Repo)
+
+	// Commit files
+	if err := repo.UserCommit(commitMessage); err != nil {
+		return fmt.Errorf("creating commit in %s/%s: %w", policyOrg, policyRepo, err)
+	}
+
+	// Push to fork
+	logrus.Infof("Pushing policy commit to %s/%s", userForkOrg, userForkRepo)
+	if err := repo.PushToRemote(kgithub.UserForkName, branchname); err != nil {
+		return fmt.Errorf("pushing %s to %s/%s: %w", kgithub.UserForkName, userForkOrg, userForkRepo, err)
+	}
+
+	prBody := fmt.Sprintf(`This pull request adds the SLSA source policy for github.com/%s/%s`, opts.Owner, opts.Repo)
+
+	// Create the Pull Request
+	pr, err := gh.CreatePullRequest(
+		policyOrg, policyRepo, "main",
```

Comment:
Should this use the branch from the CLI or do we just want to support main for now?

Reply:
This one is the pull request for the policy so it always goes to main